### DISCUSSION TITLE
API Request body is an empty object

### DISCUSSION THREAD

**Discussion by denprud** [vercel/community#4118] [Help]
### Summary

Hello,

For the past few days, I've been dealing with an issue where every time I try to do a Patch fetch call to a serverless function in my project, the body that I send ends up being an empty object. I've done just about everything to try to fix this, including fixing my headers for the fetch call. The strange thing is that this fetch call doesn't even work on Postman either.  Here is some code related to the problem:

### Steps to Reproduce

Serverless function from _app/api/route.js_
```javascript
import fetch from 'node-fetch'
import { NextRequest , NextResponse  } from 'next/server';

export async function PATCH(request) {
  const body  = request.body
  const url = `${process.env.NEXT_PUBLIC_DATA_API_URL}/api/update/recent`
  const encodedURL = encodeURI(url)
  const dataStream = await fetch(encodedURL, {method: 'PATCH', body: body})
  const dataJSON = await dataStream.json()
  return NextResponse.json(
    {
       body: body,
      path: request.nextUrl.pathname,
      query: request.nextUrl.search,
      cookies: request.cookies.getAll(),
    },
    {
      status: 200,
    },
  );

``` 

Fetch call from a client component
```javascript
let results = await fetch(`/api/serverless-example`,{ 
        headers: {
                        "content-type": "application/json",
                        "Accept": "application/json"
        },
        method: 'PATCH',
        mode: "cors",
        body: JSON.stringify({test:6})
      }).then(resp => {
             return resp.json()
        }).then(data => {
          console.log(data)
        });
``` 

vercel.json
```
{
  "headers": [
    {
      "source": "/api/(.*)",
      "headers": [
        { "key": "Access-Control-Allow-Credentials", "value": "true" },
        { "key": "Access-Control-Allow-Origin", "value": "*" },
        { "key": "Access-Control-Allow-Methods", "value": "GET,OPTIONS,PATCH,DELETE,POST,PUT" },
        { "key": "Access-Control-Allow-Headers", "value": "X-CSRF-Token, X-Requested-With, Accept, Accept-Version, Content-Length, Content-MD5, Content-Type, Date, X-Api-Version" }
      ]
    }
  ]
}
``` 

Result:
![Screenshot 2023-09-17 122853](https://github.com/vercel/community/assets/87039269/2087949a-e206-4c42-b2f3-a28af7b67fc4)

I keep getting an empty body. Any help with this is appreciated, thanks.

**Comment by jasondainter** [comment]
Did you ever solve this? I am having a similar issue!

**Comment by najwer23** [comment]
Any updates?

**Comment by amyegan** [comment]
Try changing `const body  = request.body` to `const body = await request.json();`. You might also want to set CORS headers in `next.config.js` instead of `vercel.json`.

- [Route Handlers: Request Body](https://nextjs.org/docs/app/building-your-application/routing/route-handlers#request-body)
- [next.config.js: CORS headers](https://nextjs.org/docs/app/api-reference/next-config-js/headers#cors)

**Comment by najwer23** [comment]
But I'm doing serverless with Fastify on Vercel :( 

It works with this version of Vercel:

// Vercel CLI 31.1.1, npm i -g vercel@31.1
// Vercel CLI 32.0.1, npm i -g vercel@32.0.1

But it doesnt with latest. 34

There is definitely something wrong on Vercel's side

**Reply by amyegan** [reply]
Do you have a [minimal reproducible example](https://vercel.com/support/articles/creating-a-minimal-reproducible-example) I can use to investigate? My example app works with CLI v34.1.9

**Reply by najwer23** [reply]
Sure thing Amy. I use Windows 11 and Google Chrome latest version.

1) I have app hosted on github pages (React App):
    Web: https://najwer23.github.io
    Repo: https://github.com/najwer23/najwer23.github.io
    Dev: http://localhost:3000
    
    And I have serverless App with Vercel background (Fastify, Vercel, TypeScript, ESM modules)
2) 
   Web: https://amy-from-vercel.vercel.app/static/index.html 
   Repo: https://github.com/najwer23/amy-from-vercel
   Dev: http://localhost:5000
   
3) Steps to reproduce 

CASE 1:  Vercel CLI v32.0.1 ---- It works fine

1) I start my React App: http://localhost:3000 (It's really just about the localhost address with port 3000)
2) I start Vercel App: http://localhost:5000
3) I send POST request from my react app (from devtools). 
```
fetch("http://localhost:5000" + "/mail/contact-me", {
    method: "POST",
    credentials: "include",
    headers: {
        "Content-Type": "application/json",
        "Accept": "application/json",
    },
    body: JSON.stringify({
        email: "yennefer",
        msg: "lilac and gooseberries"
    })
});
```
And I receive a response from the server
```
{
    "statusCode": 200,
    "code": "OK",
    "message": "OK",
    "error": "",
    "msg": "lilac and gooseberries",
    "email": "yennefer"
}
```

CASE 2: Vercel CLI 34.1.9 ---- It doesnt work

Response: 

Access to fetch at 'http://localhost:5000/mail/contact-me' from origin 'http://localhost:3000' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.

Case 3: Vercel web address with swagger response and with Vercel CLI 34.1.9 ---- It works fine 

Swagger response on web https://amy-from-vercel.vercel.app/static/index.html

Case 4: External addresses with Vercel CLI 34.1.9 ---- It works fine 

https://najwer23.github.io -> https://amy-from-vercel.vercel.app

Case 5. Postman  ---- It works fine 

It works beacuse Postamn doesnt care about CORS

Summary

- TBH I think you are overwriting the "headers" from my React app request
- Fetch request also doesn't work with other headers (like OPTIONS)

I hope it helps.
